src/Middleware/EnsureJwtBelongsToServiceAccountWithSyncRole.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace KeycloakAuthGuard\Middleware;
+
+use Closure;
+use Illuminate\Auth\Access\AuthorizationException;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Config;
+use KeycloakAuthGuard\Services\Decoders\RequestBasedJwtTokenDecoder;
+use stdClass;
+use Symfony\Component\HttpFoundation\Response;
+
+readonly class EnsureJwtBelongsToServiceAccountWithSyncRole
+{
+    public function __construct(private RequestBasedJwtTokenDecoder $jwtDecoder)
+    {
+    }
+
+    /**
+     * @throws AuthorizationException
+     */
+    public function handle(Request $request, Closure $next): Response
+    {
+        if (! $this->isAuthorized()) {
+            throw new AuthorizationException("You don't have the corresponding role to perform the action", 403);
+        }
+
+        return $next($request);
+    }
+
+    private function isAuthorized(): bool
+    {
+        $decodedJwt = $this->jwtDecoder->getDecodedJwtWithSpecifiedValidation(false, true);
+
+        abort_if(empty($decodedJwt), 401);
+
+        return $this->hasValidServiceAccountAzp($decodedJwt) && $this->hasSyncRole($decodedJwt);
+    }
```

Comment:
**[not-up-to-me-but-might-be-blocker]** If I remember correctly, @thenouan took some issue with authorized party (azp) validation and was thinking about removing it. I personally have no issue with it, but you might want to sync it with him.

Reply:
The issue was that the azp restricts which access tokens the service should accept. Currently we have fixed number of services (/clients, each service will receive their own oauth client) and it is easy to configure a list of azp values that service can accept. But in the near future this wouldn't help anymore when we need to allow programmatic access of some external 3rd services into our APIs. For such 3rd party services a oauth client also needs to be created, which would not match the configured list of azp values. Better solution to restrict such accesses are through oauth scopes or having programmatic roles in keycloak's side